File: BOJ/10000~19999/14889_스타트와_링크/14889_스타트와_링크.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>
using namespace std;

int n;
int ans = 987654321;
vector<vector<int> > v;

void dfs(int loc, int c, int bit)
{
  if(c == n/2){
    int a = 0;
    int b = 0;
    for(int i=0; i<n; ++i){
      for(int j=i+1; j<n; ++j){
        if((1<<i & bit) == 0){
          if((1<<j & bit) == 0){
            a += v[i][j];
            a += v[j][i];
          }
        } else {
          if((1<<j & bit) != 0){
            b += v[i][j];
            b += v[j][i];
          }
        }        
      }
    }
    ans = min(ans, abs(a-b));
    return;
  }
  for(int i=loc; i<n; ++i){
    if((1<<i & bit) == 0){
      dfs(i, c+1, (bit | 1<<i));
    }
  }
}

int main(int argc, char *argv[])
{
  int bit;
  
  cin >> n;

  v.resize(n, vector<int>(n,0));

  for(int i=0; i<n; ++i){
    for(int j=0; j<n; ++j){
      cin >> v[i][j];
    }
  }

  for(int i=0; i<n; ++i){
    dfs(i, 1, 1<<i);
  }

  cout << ans << endl;

  return 0;
}

File: BOJ/10000~19999/14889_스타트와_링크/14889_스타트와_링크.java
import java.util.*;
import java.io.*;
 
public class Main {
  static int n;
  static int[][] arr;
  static boolean[] team;
  static int ans = Integer.MAX_VALUE;
    
  public static void main(String args[]) throws Exception {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
      
    n = Integer.parseInt(br.readLine());
      
    arr = new int[n][n];
    team = new boolean[n];
      
    for(int i=0; i<n; ++i){
      StringTokenizer str = new StringTokenizer(br.readLine());
      for(int j=0; j<n; ++j){
        arr[i][j] = Integer.parseInt(str.nextToken());
      }
    }
      
    dfs(0, 0);
 
    System.out.println(ans);
  }
  static void dfs(int index, int cnt) {
    if(n / 2 == cnt){
      diff();
      return;
    }
    
    for(int i=index; i<n; ++i){
      team[i] = true;
      dfs(i+1, cnt+1);
      team[i] = false;
    }
  }
  
  static void diff() {
    int a = 0;
    int b = 0;
  
    for(int i=0; i<n-1; ++i){
      for(int j=i+1; j<n; ++j){
        if(team[i] == true && team[j] == true){
          a += arr[i][j];
          a += arr[j][i];
        } else if (team[i] == false && team[j] == false){
          b += arr[i][j];
          b += arr[j][i];
        }
      }
    }
    ans = Math.min(ans, Math.abs(a-b));
  }
}